### DISCUSSION TITLE
How to make a model provider in 1.21.4

### DISCUSSION THREAD

**Discussion by JCS-Mecabricks** [FabricMC/fabric-api#4273] [Mod Dev Support]
### Description

I updated my mods to 1.21.4 but fabric model provider doesn't exist so what fo I use?

**Comment by cputnam-a11y** [comment]
In 1.21.4, model datagen has been moved to the client sourceset, I believe there is documentation in a fabric blog post about the changes you can make to your build.gradle to convert

**Reply by JCS-Mecabricks** [reply]
Not about the model provider though. Thanks for the info.

**Reply by cputnam-a11y** [reply]
If you say so 😄

**Comment by JCS-Mecabricks** [comment]
Could you make a class of it?

**Reply by Linguardium** [reply]
https://maven.fabricmc.net/docs/fabric-api-0.110.5+1.21.4/net/fabricmc/fabric/api/client/datagen/v1/provider/FabricModelProvider.html

yes

**Comment by JCS-Mecabricks** [comment]
Can you write an example down? I would appreciate it if you could.

**Reply by Linguardium** [reply]
https://wiki.fabricmc.net/tutorial:datagen_model

**Comment by JCS-Mecabricks** [comment]
Also, I thought Fabric Model Provider was removed.

**Reply by cputnam-a11y** [reply]
Just moved

**Reply by JCS-Mecabricks** [reply]
Thank you. Do I implement or extend?

**Reply by JCS-Mecabricks** [reply]
Also, when I type Fabric Model Provider, there isn't a class option.

**Comment by JCS-Mecabricks** [comment]
Doesn't work.

**Reply by Linguardium** [reply]
Works fine

**Comment by JCS-Mecabricks** [comment]
But Model Provider isn't a class here is a picture: 
![Screenshot 2024-12-07 070553](https://github.com/user-attachments/assets/02201ede-2953-4bce-811e-81f7a9f0da05)

**Reply by rafaelmehdiyev** [reply]
Have you able to solve this ? Because of that i can't run data generation

**Reply by Linguardium** [reply]
> Have you able to solve this ? Because of that i can't run data generation

Is your data generation in your client sourceset? Model providers are clientside

**Reply by JCS-Mecabricks** [reply]
My mod was from 1.21.3 so datagen was in the main package.

**Comment by Yang00002** [comment]
I find it's able to make model generator available just by changing the `build.gradle`

from

```
fabricApi {
    configureDataGeneration()
}
```

to

```
loom {
    runs {
        datagen {
            client()

            name "Data Generation"
            vmArg "-Dfabric-api.datagen"
            vmArg "-Dfabric-api.datagen.output-dir=${file('src/main/generated')}"

            runDir "build/datagen"
        }
    }
}

sourceSets { main { resources { srcDirs += 'src/main/generated' } } } // adding generated directory to sourceset
```

and putting datagen to `client` sourceSet.

The Fabric Document said I can use
```
fabricApi {
	configureDataGeneration() {
		client = true
	}
}
```

but it doesn't work for me.

**Reply by Linguardium** [reply]
This does require loom 1.9

**Reply by dombomb64** [reply]
Neither of these get me past the error: `java.lang.RuntimeException: Cannot load class net.fabricmc.fabric.api.datagen.v1.provider.FabricModelProvider in environment type SERVER`. Mine is updated from a slightly older version of the game, if that helps.

Edit: I fixed it by going into the run configuration and replacing every instance of "server" with "client".

**Comment by EntityPlantt** [comment]
For everyone that got to the end of this post and still has issues:
- split the datagen entry point into 2 classes: for models of block/item and other stuff
- the first entry point goes in `/src/client`, whereas the second one goes in `/src/main`
- ModelProvider goes in `/src/client`, whereas other stuff remains in `/src/main`
- Add both entrypoints in `fabric.mod.json`
I'm surprised that it doesn't throw because `DatagenClient` isn't in the `main` scope and `DatagenMain` isn't in the `client`

**Reply by EntityPlantt** [reply]
Nevermind, putting `DatagenClient` after `DatagenMain` erases all generated files by `DatagenMain`

Anyone has a solution for this?

**Reply by cputnam-a11y** [reply]
Fabric only supports one active data generator. Remove the main one.

**Reply by EntityPlantt** [reply]
Thank you! I didn't know that by putting recipe, loot table and tag generators in `client`, the output remains in `/src/main/generated`
